src/lib/operators/table_scan.cpp

```diff
@@ -0,0 +1,151 @@
+        });
+      } else if (const auto reference_segment = std::dynamic_pointer_cast<ReferenceSegment>(segment)) {
+        // When handling a reference segement in a double scan, we need to refer back to the original table
+        resolve_comparator<Type>(_scan_type, [&](auto comparator) {
+          input_table = scan_reference_segment<Type>(reference_segment, search_value, comparator, position_list);
+        });
+        break;
+      } else {
+        Fail("Cannot cast the segment for TableScan");
+      }
+    }
+    position_list->shrink_to_fit();
+  });
+
+  // Save a refererence segment with the same position list for all columns
+  // We only create one output chunk
+  for (ColumnID column_id{0}; column_id < column_count; ++column_id) {
+    output_table->add_column_definition(input_table->column_name(column_id), input_table->column_type(column_id));
+    output_chunk->add_segment(std::make_shared<ReferenceSegment>(input_table, column_id, position_list));
+  }
+
+  return output_table;
+}
+
+template <typename T, typename Comparator>
+void TableScan::scan_value_segment(const std::shared_ptr<ValueSegment<T>> segment, const T search_value,
+                                   Comparator comparator, const ChunkID chunk_id,
+                                   std::shared_ptr<PosList> position_list) {
+  // Loop throught all the values and check if we match the search value. If yes, we can save the position
+  auto const values = segment->values();
+  for (ChunkOffset chunk_offset{0}; chunk_offset < segment->size(); ++chunk_offset) {
+    bool should_emit = comparator(values[chunk_offset], search_value);
+    if (should_emit) {
+      position_list->emplace_back(RowID{chunk_id, chunk_offset});
+    }
+  }
+}
+
+template <typename T, typename Comparator>
+void TableScan::scan_dictionary_segment(std::shared_ptr<DictionarySegment<T>> segment, const T search_value,
+                                        Comparator comparator, const ChunkID chunk_id,
+                                        std::shared_ptr<PosList>& position_list) {
+  if ((segment->lower_bound(search_value) == INVALID_VALUE_ID &&
+       (_scan_type == ScanType::OpGreaterThanEquals || _scan_type == ScanType::OpEquals)) ||
+      (segment->upper_bound(search_value) == INVALID_VALUE_ID && _scan_type == ScanType::OpGreaterThan)) {
+    // 1) There is no value >= search_value in the dictionary, and predicate is one of
+    // OpGreaterThanEquals, or OpEquals.
+    // OR
+    // 2) There is no value > search_value, and predicate is OpGreaterThan.
+    // Hence, we can return.
+    return;
+  }
+
+  // For dictionary segments we used the order property of ValueID. We compare the ValueId of our search value
+  // instead of comparing the values directly
+  for (ChunkOffset chunk_offset{0}; chunk_offset < segment->size(); ++chunk_offset) {
+    bool should_emit = comparator(segment->get(chunk_offset), search_value);
```

Comment:
You could get the ValueID of the search value once and compare this ValueID to the entries of the attribute vector.